sw/host/spiflash/updater.cc

```diff
@@ -77,31 +68,16 @@ bool Updater::Run() {
               << std::setw(8) << std::hex << f.hdr.offset << std::endl;
 
     if (!spi_->TransmitFrame(reinterpret_cast<const uint8_t *>(&f),
-                             reinterpret_cast<uint8_t *>(&ack[0]),
                              sizeof(Frame))) {
       std::cerr << "Failed to transmit frame no: 0x" << std::setfill('0')
                 << std::setw(8) << std::hex << f.hdr.frame_num << std::endl;
     }
 
-    // When we send the next frame, we'll get the previous frame's hash as
-    // the ack, so with each increment of |current_frame| we also need to
-    // update the |ack_expected| value.
-    uint32_t ack_expected_index = 0;
-    if (current_frame == 0 || CheckAckHash(ack, ack_expected)) {
-      ack_expected_index = current_frame;
+    // When we send each frame we wait for the correct hash before continuing.
+    if (current_frame == frames.size() - 1 ||
```

Comment:
since now you are blocking processing of the next frame, maybe this code could be refactored to something like. wdyt?

```c++
for (const Frame &f : frames) {
   TransmitFrame(...);
   BlockUntil(CheckHash(...));
}
```

Reply:
Are you suggesting a refactor where `CheckHash()` returns after reading the hash once and moving the retry and timeout logic into `updater.cc`? `TransmitFrame` used to block, so really I just split up `TransmitFrame` into two parts and moved the blocking part into `CheckHash`. I was trying to avoid going too crazy on the refactoring, but I agree it's a little awkward this way. Maybe just renaming `CheckHash` to `WaitForHash`?

Reply:
I don't believe the last frame is ever checked. AFAICT the device jumps into flash after receiving the last frame, so I don't think the last frame is ever checked. In my testing the hash received on the last frame does not match the computed hash.